packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -465,11 +466,15 @@ function validateClassAttr(vnode: VBaseElement, elm: Element, renderer: Renderer
             }
         }
 
-        vnodeClassName = computedClassName.trim();
+        readableVnodeClassname = computedClassName.trim();
 
         if (classList.length > keys(classMap).length) {
             nodesAreCompatible = false;
         }
+    } else if (elmClassName !== '') {
+        // SSR contains a className but client-side VDOM does not
+        nodesAreCompatible = false;
+        readableVnodeClassname = '';
     }
```

Comment:
The above code block adds the missing check. If a classname was present in the SSR'd HTML but not in client-side VDOM (at the time of validation), validation would succeed when it shouldn't.

Here, we ensure that, if the element has a non-empty classname, the VDOM has a non-empty classname too. If that doesn't hold true, validation now fails.